Doracleupdate (#58)
iExecBlockchainComputing/documentation: use-cases/iexec-doracle.md

**Off-chain component**

> The off-chain part of a dOracle is a classical iExec dApp, that will be executed on the iExec platform and be replicated on several workers as part of an iExec computation deal. It contains the oracle logic, for example to query a web API and process the result. Whenever an operator wishes to update the dOracle, it requests a computation like in a normal iExec deal, specifying the dOracle app as dApp, and the parameters if applicable. The dOracle result is written in the `${IEXEC_OUT}/computed.json` file by the dApp, under the `callback_data` name.

Reviewer: under the `callback_data` name.
```suggestion
The off-chain part of a dOracle is a classical iExec dApp, that will be executed on the iExec platform and be replicated on several workers as part of an iExec computation deal. It contains the oracle logic, for example to query a web API and process the result. Whenever an operator wishes to update the dOracle, it requests a computation like in a normal iExec deal, specifying the dOracle app as dApp, and the parameters if applicable. The dOracle result is written in the `${IEXEC_OUT}/computed.json` file by the dApp, under the `callback_data` key.
```

### The PriceFeed dApp

The PriceFeed dApp is a simple Node.js script, available at [Kaiko PriceFeed Source](https://github.com/iExecBlockchainComputing/iexec-apps/blob/master/offchain-computing/offchain-tee-kaiko-pricefeed/src/app.py). Given a set of parameters, the application encodes its result so that it can be interpreted by the corresponding dOracle smart contract, stores it in `${IEXEC_OUT}/computed.json`, and stores the hash of this encoded value to perform the consensus. The Worker will then send these values on-chain as part of the task finalization, where they will be accessible by the dOracle smart contract.

For example, given the parameters `"BTC USD 9 2019-04-11T13:08:32.605Z"` the price-oracle application will:

* Retrieve the price of BTC in USD at 2019-04-11T13:08:32.605Z
* Multiply this value by `10e9` \(to capture the price value more accurately as it will be represented by an integer onchain\)
* Encode the date, the description \(`"btc-usd-9"`\) and the value using `abi.encode`

> * Store this result in `${IEXEC_OUT}/computed.json` under the `callback-data` name

Reviewer: ```suggestion
* Store this result in `${IEXEC_OUT}/computed.json` under the `callback-data` key
```

internal
```

> The update function, that takes in input a task id, and reads the `Task` object data from the `IexecProxy` smart contract to perform the required checks: that the execution is complete and that the authorized app, dataset, workerpool, trust level and tags are valid. The `IexecProxy` already checked that the hash of the `resultsCallback` is equal to the `resultDigest` \(over which the consensus was reached\). If the task passes the checks then it returns the `results` field of the `Task` object, i.e. the result of the dOracle dApp computation.

Reviewer: ```suggestion
The update function, that takes in input a task id, and reads the `Task` object data from the `IexecProxy` smart contract to perform the required checks (the execution must be completed; the app, the dataset, and the workerpool must be authorized; the trust level and tags mus be valid). The `IexecProxy` already checked that the hash of the `resultsCallback` is equal to the `resultDigest` \(over which the consensus was reached\). If the task passes the checks then it returns the `results` field of the `Task` object, i.e. the result of the dOracle dApp computation.
```
